libminifi/src/core/state/nodes/ResponseNodeLoader.cpp

```diff
@@ -25,89 +25,129 @@
     component_metrics_.clear();
   }
 
-  if (!root) {
+  std::lock_guard<std::mutex> guard(root_mutex_);
+  if (!root_) {
     return;
   }
 
   std::vector<core::Processor*> processors;
-  root->getAllProcessors(processors);
+  root_->getAllProcessors(processors);
   for (const auto processor : processors) {
     auto node_source = dynamic_cast<ResponseNodeSource*>(processor);
     if (node_source == nullptr) {
       continue;
     }
     // we have a metrics source.
-    auto metric = node_source->getResponseNodes();
+    auto metric = node_source->getResponseNode();
     std::lock_guard<std::mutex> guard(component_metrics_mutex_);
     component_metrics_[metric->getName()].push_back(metric);
   }
 }
 
-std::vector<std::shared_ptr<ResponseNode>> ResponseNodeLoader::getResponseNodes(const std::string& clazz) const {
-  std::shared_ptr<core::CoreComponent> ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
-  if (ptr == nullptr) {
-    return getComponentMetricsNodes(clazz);
+nonstd::expected<SharedResponseNode, std::string> ResponseNodeLoader::getSystemMetricsNode(const std::string& clazz) {
+  std::lock_guard<std::mutex> guard(system_metrics_mutex_);
+  if (system_metrics_.contains(clazz)) {
+    return system_metrics_.at(clazz);
   }
+
+  std::shared_ptr ptr = core::ClassLoader::getDefaultClassLoader().instantiate(clazz, clazz);
   auto response_node = std::dynamic_pointer_cast<ResponseNode>(ptr);
   if (!response_node) {
-    logger_->log_error("Instantiated class '%s' is not a ResponseNode!", clazz);
+    return nonstd::make_unexpected("Instantiated class '" + clazz + "' is not a ResponseNode!");
+  }
+  system_metrics_.emplace(clazz, gsl::make_not_null(response_node));
+  return system_metrics_.at(clazz);
+}
+
+std::vector<SharedResponseNode> ResponseNodeLoader::getResponseNodes(const std::string& clazz) {
+  auto component_metrics = getComponentMetricsNodes(clazz);
+  if (!component_metrics.empty()) {
+    return component_metrics;
+  }
+  auto response_node = getSystemMetricsNode(clazz);
+  if (!response_node) {
+    logger_->log_error(response_node.error().c_str());
     return {};
   }
-  return {response_node};
+  return {*response_node};
 }
 
-void ResponseNodeLoader::initializeRepositoryMetrics(const std::shared_ptr<ResponseNode>& response_node) const {
+void ResponseNodeLoader::initializeRepositoryMetrics(const SharedResponseNode& response_node) const {
   auto repository_metrics = dynamic_cast<RepositoryMetrics*>(response_node.get());
   if (repository_metrics != nullptr) {
     repository_metrics->addRepository(provenance_repo_);
     repository_metrics->addRepository(flow_file_repo_);
   }
 }
 
-void ResponseNodeLoader::initializeQueueMetrics(const std::shared_ptr<ResponseNode>& response_node, core::ProcessGroup* root) {
-  if (!root) {
+void ResponseNodeLoader::initializeQueueMetrics(const SharedResponseNode& response_node) const {
+  std::lock_guard<std::mutex> guard(root_mutex_);
```

Comment:
I would reduce the scope of this lock to just the `getConnections()` call, too.

Reply:
This should be kept due to the same issue as above.